lib/ddtrace/context.rb

```diff
@@ -200,19 +201,23 @@ def check_finished_spans
     end
 
     def attach_sampling_priority
-      @trace.first.set_metric(
+      @current_root_span.set_metric(
         Ext::DistributedTracing::SAMPLING_PRIORITY_KEY,
         @sampling_priority
       )
     end
 
     def attach_origin
-      @trace.first.set_tag(
+      @current_root_span.set_tag(
         Ext::DistributedTracing::ORIGIN_KEY,
         @origin
       )
     end
 
+    def associate_runtime_metrics
+      Datadog.runtime_metrics.associate_with_span(@current_root_span)
```

Comment:
Not sure I like that `Context` now directly depends on some global level `Datadog.runtime_metrics` reference. Context should be more self-contained than this.

Should `runtime_metrics` be an attribute of the `Worker` instead? If so, we'd remove this here, then in the `Writer` call `worker.runtime_metrics.associate_with_span(root_span)`. (Or something like that.)

Reply:
You know this tracer much better than I do, especially where some boundaries are and should be.

I think the only thing to really keep in mind is we probably want the runtime metrics collector to run in it's own thread/process in the future, rather than being tied directly to the Writer.

So, I think given that suggestion, it makes sense to add runtime metrics as a property on `Writer`, because even if it is "hard coded" into the writer today, nothing is stopping us from passing it in as a constructor variable in the future if we needed to decouple them.

Reply:
Yeah, I think it's a step in the right direction, and sets us up for extracting runtime metrics into its own process.

If you think about it, `Writer` doesn't really write... it's more of a "processor" that glues a trace buffer and output (`Transport`) to a thread (`Worker`.) Eventually we should have a component that parallels this behavior but for runtime metrics.

I have some grander ideas of how to reorganize this later, but for now, it probably should piggyback on the `Writer`.

Reply:
:+1: